src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,208 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+
+    private int size;
+    private Node<T> first;
+    private Node<T> last;
+
+    public MyLinkedList() {
+        size = 0;
+    }
+
     @Override
     public boolean add(T value) {
-        return false;
+        linkLast(value);
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index == size) {
+            linkLast(value);
+        } else {
+            linkBefore(value, node(index));
+        }
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        return addAll(list, size);
+    }
+
+    public boolean addAll(List<T> list, int index) {
+        Object[] a = list.toArray();
+        int numNew = a.length;
+        if (numNew == 0) {
+            return false;
+        }
+
+        Node<T> pred;
+        Node<T> succ;
+        if (index == size) {
+            succ = null;
+            pred = last;
+        } else {
+            succ = node(index);
+            pred = succ.prev;
+        }
+
+        for (Object o : a) {
+            Node<T> newNode = new Node<T>(pred, (T) o, null);
+            if (pred == null) {
+                first = newNode;
+            } else {
+                pred.next = newNode;
+            }
+            pred = newNode;
+        }
+
+        if (succ == null) {
+            last = pred;
+        } else {
+            pred.next = succ;
+            succ.prev = pred;
+        }
+
+        size += numNew;
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return node(index).item;
     }
 
     @Override
-    public T set(T value, int index) {
-        return null;
+    public T set(T element, int index) {
+        Node<T> x = node(index);
+        T oldVal = x.item;
+        x.item = element;
+        return oldVal;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> node = node(index);
+        unlink(node);
+        return node.item;
     }
 
     @Override
     public boolean remove(T t) {
+        if (t == null) {
```

Comment:
do we need this checking? let's use only one for loop

Reply:
done